### PR TITLE
Remove internal build logic from public build tool plugins

### PR DESCRIPTION
**Pull request by breskeby**
Extract usage of internal API from TestClustersPlugin and PluginBuildPlugin and related plugins and build logic

This includes a refactoring of ElasticsearchDistribution to handle types
better in a way we can differentiate between supported Elasticsearch
Distribution types supported in TestCkustersPlugin and types only supported
in internal plugins.
 
It also introduces a set of internal versions of public plugins.

As part of this we also generate the plugin descriptors now.

As a follow up on this we can actually move these public used classes into an extra project (declared as included build)

### CODE REVIEW

File: buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java [hunk 1]
```diff
@@ -48,14 +46,19 @@
...
     private NamedDomainObjectContainer<DistributionResolution> distributionsResolutionStrategiesContainer;
 
+    private Property<Boolean> dockerAvailability;
+
+    @Inject
+    public DistributionDownloadPlugin(ObjectFactory objectFactory) {
+        this.dockerAvailability = objectFactory.property(Boolean.class).value(false);
+    }
+
+    public void setDockerAvailability(Provider<Boolean> dockerAvailability) {
```
Comment (breskeby) [inline buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java:56]: Abstracting away the (only internally available) optional DockerSupportService

File: buildSrc/src/main/java/org/elasticsearch/gradle/ElasticsearchDistribution.java [hunk 2]
```diff
@@ -36,49 +34,6 @@ public String toString() {
         }
     }
 
-    public enum Type {
```
Comment (breskeby) [inline buildSrc/src/main/java/org/elasticsearch/gradle/ElasticsearchDistribution.java:39]: We replace this Type enum with a proper interface. 
That way we can have proper public and internal Types and hide the Docker based distributions that are not supported by the public `TestClustersPlugin` and is only used in our internal builds 
Also switches with defaults are evil and we should avoid them IMO

File: buildSrc/src/main/java/org/elasticsearch/gradle/ReaperPlugin.java [hunk 3]
```diff
@@ -32,20 +30,24 @@ public ReaperPlugin(ProjectLayout projectLayout) {
...
             .dir(".gradle")
             .dir("reaper")
             .dir("build-" + ProcessHandle.current().pid())
             .getAsFile();
+
         project.getGradle().getSharedServices().registerIfAbsent(REAPER_SERVICE_NAME, ReaperService.class, spec -> {
             // Provide some parameters
             spec.getParameters().getInputDir().set(inputDir);
             spec.getParameters().getBuildDir().set(projectLayout.getBuildDirectory());
-            spec.getParameters().setInternal(BuildParams.isInternal());
```
Comment (breskeby) [inline buildSrc/src/main/java/org/elasticsearch/gradle/ReaperPlugin.java:48]: Remove internal BuildParams usage as external usage is never internal anyway.

File: buildSrc/src/main/java/org/elasticsearch/gradle/internal/BuildPlugin.java [hunk 4]
```diff
@@ -44,18 +41,7 @@ public void apply(final Project project) {
...
-
-    private static void checkExternalInternalPluginUsages(Project project) {
-        if (BuildParams.isInternal().equals(false)) {
-            project.getPlugins()
-                .withType(
-                    InternalPlugin.class,
-                    internalPlugin -> { throw new GradleException(internalPlugin.getExternalUseErrorMessage()); }
-                );
-        }
+        InternalPrecommitTasks.create(project, true);
```
Comment (breskeby) [inline buildSrc/src/main/java/org/elasticsearch/gradle/internal/BuildPlugin.java:44]: This plugin is now only applied internally

File: buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginBuildPlugin.java [hunk 5]
```diff
@@ -112,54 +107,7 @@ public void apply(final Project project) {
                 copy.expand(map);
                 copy.getInputs().properties(map);
             });
-            BuildParams.withInternalBuild(() -> {
```
Comment (breskeby) [inline buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginBuildPlugin.java:115]: All internal specific logic moved into `InternalPluginBuildPlugin`

File: buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginBuildPlugin.java [hunk 6]
```diff
@@ -193,13 +141,8 @@ private static void configurePublishing(Project project, PluginPropertiesExtensi
...
     private static void configureDependencies(final Project project) {
         var dependencies = project.getDependencies();
-        if (BuildParams.isInternal()) {
-            dependencies.add("compileOnly", dependencies.project(Map.of("path", ":server")));
-            dependencies.add("testImplementation", dependencies.project(Map.of("path", ":test:framework")));
-        } else {
-            dependencies.add("compileOnly", "org.elasticsearch:elasticsearch:" + VersionProperties.getElasticsearch());
-            dependencies.add("testImplementation", "org.elasticsearch.test:framework:" + VersionProperties.getElasticsearch());
-        }
+        dependencies.add("compileOnly", "org.elasticsearch:elasticsearch:" + VersionProperties.getElasticsearch());
```
Comment (breskeby) [inline buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginBuildPlugin.java:144]: We handle this differently and substitute it in our internal build to avoid using internal `BuildParams`

File: buildSrc/src/main/java/org/elasticsearch/gradle/precommit/JarHellPrecommitPlugin.java [hunk 7]
```diff
@@ -20,12 +19,6 @@
     @Override
     public TaskProvider<? extends Task> createTask(Project project) {
         Configuration jarHellConfig = project.getConfigurations().create("jarHell");
-        if (BuildParams.isInternal() && project.getPath().equals(":libs:elasticsearch-core") == false) {
```
Comment (breskeby) [inline buildSrc/src/main/java/org/elasticsearch/gradle/precommit/JarHellPrecommitPlugin.java:23]: We moved setting the dependencies here to `InternalPrecommitTask`. to not rely on internal BuildParams and keep this public Plugin cleaner with none internal specific polluted code. We could alternatively have an InternalJarHellPrecommitPlugin but as we want to revisit default dependency declarations in our plugins and our internal build that's likely too much.

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java [hunk 8]
```diff
@@ -762,7 +765,7 @@ private void runElasticsearchBinScript(String tool, CharSequence... args) {
         Map<String, String> defaultEnv = new HashMap<>();
         // If we are testing the current version of Elasticsearch, use the configured runtime Java, otherwise use the bundled JDK
         if (getTestDistribution() == TestDistribution.INTEG_TEST || getVersion().equals(VersionProperties.getElasticsearchVersion())) {
-            defaultEnv.put("ES_JAVA_HOME", BuildParams.getRuntimeJavaHome().getAbsolutePath());
```
Comment (breskeby) [inline buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java:765]: Remove last internal class usage from Public ElasticsearchNode

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java [hunk 9]
```diff
@@ -58,16 +59,26 @@ protected ExecOperations getExecOperations() {
...
-        project.getRootProject().getPluginManager().apply(GlobalBuildInfoPlugin.class);
-        BuildParams.withInternalBuild(() -> project.getPlugins().apply(InternalDistributionDownloadPlugin.class))
-            .orElse(() -> project.getPlugins().apply(DistributionDownloadPlugin.class));
+        project.getPlugins().apply(DistributionDownloadPlugin.class);
         project.getRootProject().getPluginManager().apply(ReaperPlugin.class);
         Provider<ReaperService> reaperServiceProvider = GradleUtils.getBuildService(
             project.getGradle().getSharedServices(),
             ReaperPlugin.REAPER_SERVICE_NAME
         );
+        runtimeJavaProvider = providerFactory.provider(
```
Comment (breskeby) [inline buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java:79]: for public usage we just rely on a simple mechanism for declaring the JAVA RUNTIME

File: buildSrc/src/main/java/org/elasticsearch/gradle/internal/InternalPluginBuildPlugin.java [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+import org.elasticsearch.gradle.internal.util.Util;
+import org.elasticsearch.gradle.plugin.PluginBuildPlugin;
+import org.elasticsearch.gradle.plugin.PluginPropertiesExtension;
+import org.elasticsearch.gradle.util.GradleUtils;
+import org.gradle.api.Project;
+import org.gradle.api.tasks.bundling.Zip;
+
+import java.util.Optional;
+
+public class InternalPluginBuildPlugin implements InternalPlugin {
```
Comment (breskeby) [inline buildSrc/src/main/java/org/elasticsearch/gradle/internal/InternalPluginBuildPlugin.java:23]: Contains the internal plugin specific configurations we used to keep in PluginBuildPlugin

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestDistribution.java [hunk 11]
```diff
@@ -13,5 +13,4 @@
 public enum TestDistribution {
     INTEG_TEST,
     DEFAULT,
-    OSS
```
Comment (breskeby) [inline buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestDistribution.java:16]: unrelated cleanup

**Comment by breskeby** [comment]
Tested with elasticsearch-hadoop: https://gradle-enterprise.elastic.co/s/bdkdwzc55liug

File: build.gradle [hunk 12]
```diff
@@ -12,7 +12,7 @@ import de.thetaphi.forbiddenapis.gradle.ForbiddenApisPlugin
 import org.apache.tools.ant.taskdefs.condition.Os
 import org.elasticsearch.gradle.internal.BuildPlugin
 import org.elasticsearch.gradle.Version
-import org.elasticsearch.gradle.internal.VersionProperties
+import org.elasticsearch.gradle.VersionProperties
```
Comment (breskeby) [inline build.gradle:15]: Public plugin still requires VersionProperties. We should revisit general version handling and support in external plugins later again

File: buildSrc/src/main/java/org/elasticsearch/gradle/internal/InternalDistributionDownloadPlugin.java [hunk 13]
```diff
@@ -149,24 +151,21 @@ private static String distributionProjectName(ElasticsearchDistribution distribu
...
-
-            case DOCKER_UBI:
-                return projectName + "ubi-docker" + archString + "-export";
-
-            case DOCKER_IRON_BANK:
-                return projectName + "ironbank-docker" + archString + "-export";
-
-            default:
-                return projectName + distribution.getType();
+        if (distribution.getType() == ElasticsearchDistributionTypes.ARCHIVE) {
```
Comment (pugnascotia) [inline buildSrc/src/main/java/org/elasticsearch/gradle/internal/InternalDistributionDownloadPlugin.java:154]: Personally I'd have left this as a `switch` because I think it's clearer, but no big deal.

Comment (pugnascotia) [reply buildSrc/src/main/java/org/elasticsearch/gradle/internal/InternalDistributionDownloadPlugin.java:154]: Ah wait, it's not an `enum` any more, I see. Ignore me 😄

File: buildSrc/src/main/java/org/elasticsearch/gradle/ElasticsearchDistributionType.java [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.gradle;
+
+public interface ElasticsearchDistributionType {
+    boolean shouldExtract();
```
Comment (pugnascotia) [inline buildSrc/src/main/java/org/elasticsearch/gradle/ElasticsearchDistributionType.java:12]: What do you think about defaulting this method and `isDocker()` to false?

File: buildSrc/src/main/java/org/elasticsearch/gradle/internal/precommit/LoggerUsagePrecommitPlugin.java [hunk 15]
```diff
@@ -19,7 +19,11 @@
     @Override
     public TaskProvider<? extends Task> createTask(Project project) {
         Configuration loggerUsageConfig = project.getConfigurations().create("loggerUsagePlugin");
-        project.getDependencies().add("loggerUsagePlugin", project.project(":test:logger-usage"));
+        // this makes it better testable by not requiring this project to be always available in our
```
Comment (pugnascotia) [inline buildSrc/src/main/java/org/elasticsearch/gradle/internal/precommit/LoggerUsagePrecommitPlugin.java:22]:
```suggestion
        // this makes it easier to test by not requiring this project to be always available in our
```

**Review by pugnascotia — APPROVED** [review]
LGTM, though there's a lot I don't know about this stuff.

File: buildSrc/src/main/java/org/elasticsearch/gradle/ElasticsearchDistributionType.java (hunk 14 above)
Comment (breskeby) [reply buildSrc/src/main/java/org/elasticsearch/gradle/ElasticsearchDistributionType.java:12]: good idea. done

File: build.gradle (hunk 12 above)
Comment (mark-vieira) [reply build.gradle:15]: Indeed. Really the only thing we use it for is to determine which version of the elasticsearch dependencies to use. We could do a few things here:

1. Just extract the version number (perhaps at build time) instead of bringing in VersionProperties and all it's friends.
2. Don't make any assumptions about ES version at all and let the plugin author specify it. That _could_ get weird if there are requirements but generally there's no reason you can't use 8.0 build-tools to build a 7.x plugin.

File: buildSrc/src/main/java/org/elasticsearch/gradle/internal/precommit/InternalPrecommitTasks.java [hunk 16]
```diff
@@ -21,7 +22,12 @@
      */
     public static void create(Project project, boolean includeDependencyLicenses) {
         PrecommitTasks.create(project);
-
+        if (BuildParams.isInternal() && project.getPath().equals(":libs:elasticsearch-core") == false) {
```
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/internal/precommit/InternalPrecommitTasks.java:25]: Why are we checking `isInternal()` here? This is an internal-only plugin so shouldn't be used by external users, right?

File: buildSrc/src/main/java/org/elasticsearch/gradle/internal/test/DistroTestPlugin.java [hunk 17]
```diff
@@ -364,7 +370,7 @@ private static void configureVMWrapperTasks(
         List<ElasticsearchDistribution> currentDistros = new ArrayList<>();
 
         for (Architecture architecture : Architecture.values()) {
-            Arrays.stream(Type.values()).filter(t -> t != Type.INTEG_TEST_ZIP && t != Type.ARCHIVE).forEach(type -> {
+            List.of(DEB, RPM, DOCKER, DOCKER_IRONBANK, DOCKER_UBI).stream().forEach(type -> {
```
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/internal/test/DistroTestPlugin.java:373]: We should probably move this list into `InternalElasticsearchDistributionTypes` that way when we add a new one we don't risk forgetting to update this.

**Review by mark-vieira — APPROVED** [review]
Sweet. LGTM.

File: build.gradle (hunk 12 above)
Comment (breskeby) [reply build.gradle:15]: I think option 2) is the way to go and we could publish a BOM as part of our releases and consume these elasticsearch related dependencies via as a platform. IMO this is the most natural way to deal with those kind of dependencies in the plugin. Again, I'd like to tackle this in a separate PR once we actually split out the build-tool project.

File: buildSrc/src/main/java/org/elasticsearch/gradle/internal/precommit/InternalPrecommitTasks.java (hunk 16 above)
Comment (breskeby) [reply buildSrc/src/main/java/org/elasticsearch/gradle/internal/precommit/InternalPrecommitTasks.java:25]: There are still some BuilParams.isInternal() usages left in our internal plugins. The main reason I didn't rip them out as part of this PR is that removing them makes a huge set of our plugin functional tests fail as they would require a more sophisticated test project structure. It's related to the problem of declaring default (project) dependencies in plugins as we discussed few days ago which makes testing our plugins way harder.

I will create some follow up tickets to track and sort this (the removal of BuildParams.isInternal()) and the better handling of dependencies out in our codebase.

File: buildSrc/src/main/java/org/elasticsearch/gradle/internal/test/DistroTestPlugin.java (hunk 17 above)
Comment (breskeby) [reply buildSrc/src/main/java/org/elasticsearch/gradle/internal/test/DistroTestPlugin.java:373]: Will do

File: build.gradle [hunk 18]
```diff
@@ -108,6 +110,22 @@ subprojects {
     project.noticeFile = project.rootProject.file('NOTICE.txt')
   }
 
+  // We substitute dependencies introduced by the PluginBuildPlugin but this won't work for projects
+  // consuming plugin projects. TODO: find a more elegant approach
+  project.getConfigurations().matching(c -> c.getName().matches("es\\d+") == false ).configureEach(c -> {
```
[the remark this answers is not in the document]
Comment (breskeby) [reply build.gradle:115]: These are definitely only temporarily as we wanna rip out the default dependency declarations from `PluginBuildPlugin`. Anyhow I will try to find an alternative to dependency substitutions to deal with this before merging.

File: buildSrc/src/main/java/org/elasticsearch/gradle/internal/precommit/InternalPrecommitTasks.java (hunk 16 above)
Comment (mark-vieira) [reply buildSrc/src/main/java/org/elasticsearch/gradle/internal/precommit/InternalPrecommitTasks.java:25]: Roger that. The test fallout makes sense.
